### PR TITLE
ci: 👷 Add action to automate go mod tidy

### LINKED ISSUE #7: Run 'go mod tidy' automatically
#### Description
Whenever there is a new version of the 3rd party dependencies used by the project (namely whatever is inside `go.mod` and `go.sum`) then let the project update itself without developer's intervention.

#### Test the first way

https://github.com/marketplace/actions/go-mod-tidy-pr-go-1-13-4

#### Test the second way

- [ ] Learn about [dependabot](https://dependabot.com/) which is available also at GitHub Marketplace [here](https://github.com/marketplace/dependabot-preview)
- [ ] Write a GH Action workflow doing the following:
- [ ] runs on `push` for the `master` branch`
- [ ] gets triggered when `go.mod`, `go.sum` or the `.github/workflow/<name>.yml` has been modified.
- [ ] fetches the project using [actions/checkout](https://github.com/actions/checkout)
- [ ] sets up Go 1.15 via [actions/setup-go](https://github.com/actions/setup-go)
- [ ] Make sure it runs this snippet:

```yml
        name: Tidy
        run: |
          rm -f go.sum
          go mod tidy
```
 
After updating those files, it prepares a PR from `go-mod-tidy` branch, that gets automatically thanks to the `automerge` labels. For creating the PR as part of the same GH action use this, with the following settings:

 - [ ] create a PR using [actions/create-pull-request](https://github.com/peter-evans/create-pull-request) with:

```yml
        name: Create Pull Request
        uses: peter-evans/create-pull-request@v2
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          commit-message: "chore(deps): go mod tidy"
          title: "chore(deps): go mod tidy"
          body: |
            Current `go.mod` and `go.sum` don't match the source code.
          branch: go-mod-tidy
          branch-suffix: timestamp
          labels: automerge
```

#### Feedback

Which of the two ways is better? Please provide pros and cons

Complexity: 8

### PR DESCRIPTION
**Pull request by zoispag**
👷 ci: Add action to automate go mod tidy

- Uses `futuratrepadeira/changed-files@v3.2.1` action to check for updates on `go.mod`, `go.sum` or `.github/workflows/go-mod-tidy.yml`
- Conditionally uses `sue445/go-mod-tidy-pr@master` to run `go mod tidy` and create a PR

Closes #7

> `pull_request` event is temporary as part of WIP to test the action

### CODE REVIEW

**Comment by zoispag** [comment]
@drpaneas if `sue445/go-mod-tidy-pr@master` is working the way it is supposed to work, then I am guessing that token provided via `secrets.GITHUB_TOKEN` does not have the proper scope to create PRs, even though the branch is created and pushed to remote.

image

image

I'll need your help here.

Once done, I will remove the `pull_request` event from this action, so it will only run on master branch and I will drop the `logrus` dependency, which was `go get` to test this action.

**Comment by zoispag** [comment]
Update: After doing some research, it looks like `secrets.GITHUB_TOKEN` is more than enough to create PRs, but will not do so, if triggered by another PR, to avoid an endless loop situation.
image

image


That being said, I will then revert `go.mod` changes and remove the `pull_request` event, to make the PR ready for merge.

**Comment by drpaneas** [comment]
Can you squash the commits?

**Comment by zoispag** [comment]
> Can you squash the commits?

@drpaneas  aren't you going to do Squash and Merge from the PR?

**Comment by drpaneas** [comment]
> > Can you squash the commits?
> 
> @drpaneas aren't you going to do Squash and Merge from the PR?

coffee ... I need ...

**Review by drpaneas — APPROVED** [review]

**Comment by zoispag** [comment]
> coffee ... I need ...

File: .github/workflows/go-mod-tidy.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+          pattern: '^(go\.(mod|sum)|\.github\/workflows\/go-mod-tidy\.yml)$'
+
+      - name: Run go-mod-tidy-pr
+        if: steps.changed_files.outputs.files_updated
+        uses: sue445/go-mod-tidy-pr@master
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          git_user_name: Dependabot
+          git_user_email: github-actions@example.com
```
Comment (cerebrux) [inline .github/workflows/go-mod-tidy.yml:31]: is this mail legit ? :P

**Review by cerebrux — CHANGES_REQUESTED** [review]

File: .github/workflows/go-mod-tidy.yml (hunk 1 above)
Comment (zoispag) [reply .github/workflows/go-mod-tidy.yml:31]: Yeap. As long as it's a valid email address, git is supposed to accept a commit.

Comment (cerebrux) [reply .github/workflows/go-mod-tidy.yml:31]: I thought it should be @drpaneas (project owner) email :)

Comment (zoispag) [reply .github/workflows/go-mod-tidy.yml:31]: I don't think a real maintainer should make arbitrary automated commits issued by bots.

Comment (zoispag) [reply .github/workflows/go-mod-tidy.yml:31]: It needs to be clear that the commit was actually originated by a **_bot_** account.

**Review by cerebrux — APPROVED** [review]

**Pull request merged**